### ISSUE TITLE
Trying to establish a testsuite, looking for grammars.

### ISSUE THREAD

**Issue by modulovalue** [jikespg/jikespg#4]
Hello everyone!

I'm currently refactoring the jikespg codebase to make it easier to translate it to a higher level language. For that I'm looking to establish a testsuite that allows me verify that my refactorings don't break anything.

In this issue I wanted to ask some people for their knowledge of any jikespg grammar files or perhaps if they could share some that they have developed themselves.

---

@daveshields since you're the original author, do you have any guidance on where I could start looking for jikespg grammars?

I also wanted to ask @btj @stephan-herrmann @xhaakon @mickaelistria, since you're known to have worked on jikespg before.

---

So far, I've only found [the eclipse java grammar here](https://wiki.eclipse.org/JDT_Core_Programmer_Guide/ECJ/Parse), the two example grammars in the example directory, and the jikespg grammar file grammar itself.

**Comment by btj** [comment]
> I'm currently refactoring the jikespg codebase to make it easier to translate it to a higher level language.

Excellent!!

As far as I know, generating the parser for Eclipse JDT is the only thing jikespg is currently being used for. (But I wouldn't really know.) In any case, as far as I'm concerned, that's the only grammar your new version needs to support.

**Comment by stephan-herrmann** [comment]
> So far, I've only found [the eclipse java grammar here](https://wiki.eclipse.org/JDT_Core_Programmer_Guide/ECJ/Parse),

Right. More specifically, the grammar is https://github.com/eclipse-jdt/eclipse.jdt.core/blob/master/org.eclipse.jdt.core.compiler.batch/grammar/java.g -- which should already exercise a significant portion of jikespg, although we haven't changed any of the global options since like forever. 

Do you have a plan how you want to test equivalence of before - after? Do you expect the exact same bytes in all of jikespg's output? Otherwise, if you are interested in an end-to-end test, we could perhaps work together on applying the full compiler test suite of Eclipse JDT on the output from the refactored jikespg.

**Comment by modulovalue** [comment]
> Do you expect the exact same bytes in all of jikespg's output?

Yes, exactly, that's the plan.

> if you are interested in an end-to-end test, we could perhaps work together on applying the full compiler test suite of Eclipse JDT on the output from the refactored jikespg.

Thank you for the offer! I guess as long as the output stays the same we wouldn't benefit from doing that.

@stephan-herrmann Do you know if the generated java parser is checked into the eclipse codebase somewhere, or is it only generated ad-hoc during some compilation step? EDIT: I've found this: https://github.com/eclipse-jdt/eclipse.jdt.core/wiki/Adding-support-for-new-Java-version-and-Features-in-JDT#parserjava

**Comment by stephan-herrmann** [comment]
> @stephan-herrmann Do you know if the generated java parser is checked into the eclipse codebase somewhere, or is it only generated ad-hoc during some compilation step? EDIT: I've found this: https://github.com/eclipse-jdt/eclipse.jdt.core/wiki/Adding-support-for-new-Java-version-and-Features-in-JDT#parserjava

This speaks of manual updates for a new Java version.

You may want to have a look at https://github.com/eclipse-jdt/eclipse.jdt.core/blob/master/org.eclipse.jdt.core.compiler.batch/scripts/GenerateParserScript.java  -- which then produces the output that indeed will be committed to git (see comments in the "script"):
* parserNN.rsc and readableNames.props
* TerminalTokens.java
* ParserBasicInformation.java
* method consumeRule in Parser.java

**Comment by modulovalue** [comment]
Thanks again @stephan-herrmann! I see there's some amount of post processing going on to mold the generated parser into a shape that eclipse expects, which I didn't expect. It might be interesting to see if java.g/jikespg can be updated to support the use case of eclipse directly without the generated files having to be post processed after generation.

---

Here are some more people that were involved in efforts related to jikespg.

@MaisiKoleni via https://github.com/MaisiKoleni/gedit
@datho7561 via https://github.com/datho7561/eclipse-jikespg
@rmfuhrer via his work on https://sourceforge.net/projects/lpg/

If you happen to read this comment, any help with grammars for an extensive test suite would be greatly appreciated!

---

I've also finally found the full LPG repo, which appears to be a rewrite of jikespg into C++: https://github.com/steleman/lpg (which is the predecessor to LPG2 (https://github.com/A-LPG/LPG2)). It appears to contain several more example grammars. I'm not sure if the specification format changed, but it looks similar, I'll have to investigate.

And I've found another jikespg use in the wild here: https://github.com/othomann/formulachess/blob/77ae9b47a6bbcd60a5f8681f4b54862cbeba96aa/bundles/org.formulachess.chess.utils/utils/pgn.g
